### PR TITLE
[6.x][WIP] Bassets 🐶

### PR DESCRIPTION
**Pull request by promatik**
Solves: https://github.com/Laravel-Backpack/community-forum/issues/116 (using Option 7)
Status: WIP (80%)

TLDR: instead of using NPM to pull the CSS and JS assets we need, then publish them, then re-publish... we point directly to assets on a CDN; the first time they're needed, they are "_internalized_" (downloaded to the app's server) so that they're actually used from there; this solves bloat issues, GDPR issues, loading time issues, inconvenience issues; 

In order to get it working you need to use the `bassets` branch on all repos. Here are their respective PRs:
- demo - https://github.com/Laravel-Backpack/demo/pull/449 <---- has installation instructions
- assets - https://github.com/DigitallyHappy/assets/pull/15
- PRO - https://github.com/Laravel-Backpack/PRO/pull/110
- CRUD - this one

TODO: In order to consider this DONE, we need to resolve all the inline comments we added here.

### CODE REVIEW

File: src/config/backpack/base.php [hunk 1]
```diff
@@ -58,8 +58,12 @@
         // 'packages/noty/noty.css',
 
         // Load the fonts separately (so that you can replace them at will):
-        'packages/source-sans-pro/source-sans-pro.css',
-        'packages/line-awesome/css/line-awesome.min.css',
+        ['https://fonts.googleapis.com/css2?family=Source+Sans+Pro:ital,wght@0,400;0,600;0,700;1,400&display=swap', true, [], 'style'],
```
Comment (tabacitu) [inline src/config/backpack/base.php:61]: This isn't working right now - loading fonts from CDNs like Google Fonts. Why? Because the URL itself doesn't have an extension... so when we save it, THAT doesn't have an extension... 

Even if it did... it would work up to a point. Inside that CSS file, here's what we've got:
```css
@font-face {
  font-family: 'Source Sans Pro';
  font-style: italic;
  font-weight: 400;
  font-display: swap;
  src: url(https://fonts.gstatic.com/s/sourcesanspro/v21/6xK1dSBYKcSV-LCoeQqfX1RYOo3qPa7g.ttf) format('truetype');
}
@font-face {
  font-family: 'Source Sans Pro';
  font-style: normal;
  font-weight: 400;
  font-display: swap;
  src: url(https://fonts.gstatic.com/s/sourcesanspro/v21/6xK3dSBYKcSV-LCoeQqfX1RYOo3aPw.ttf) format('truetype');
}
@font-face {
  font-family: 'Source Sans Pro';
  font-style: normal;
  font-weight: 600;
  font-display: swap;
  src: url(https://fonts.gstatic.com/s/sourcesanspro/v21/6xKydSBYKcSV-LCoeQqfX1RYOo3i54rAkA.ttf) format('truetype');
}
@font-face {
  font-family: 'Source Sans Pro';
  font-style: normal;
  font-weight: 700;
  font-display: swap;
  src: url(https://fonts.gstatic.com/s/sourcesanspro/v21/6xKydSBYKcSV-LCoeQqfX1RYOo3ig4vAkA.ttf) format('truetype');
}
```

Which means we'd be internalizing the CSS file, but NOT the actual font files - those would still be loaded from Google's servers. It would be working. But it wouldn't be GDPR-friendly.

File: src/resources/views/base/inc/head.blade.php [hunk 2]
```diff
@@ -11,7 +11,11 @@
 
     @if (config('backpack.base.styles') && count(config('backpack.base.styles')))
         @foreach (config('backpack.base.styles') as $path)
-        <link rel="stylesheet" type="text/css" href="{{ asset($path).'?v='.config('backpack.base.cachebusting_string') }}">
+            @if(is_array($path))
```
Comment (tabacitu) [inline src/resources/views/base/inc/head.blade.php:14]: When you're echoing multiple bassets one after another, the output is on the same line, which looks odd:

CleanShot 2023-01-03 at 21 16 26

Instead of making the source code look super-clean, it does the opposite.

File: src/resources/views/base/inc/head.blade.php [hunk 3]
```diff
@@ -27,10 +31,3 @@
 
     @yield('after_styles')
     @stack('after_styles')
-
-    {{-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries --}}
-    {{-- WARNING: Respond.js doesn't work if you view the page via file:// --}}
-    <!--[if lt IE 9]>
-    <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
-    <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
-    <![endif]-->
```
Comment (tabacitu) [inline src/resources/views/base/inc/head.blade.php:36]: Why remove the Shim and Respond.js? The way I understand it, this would remove some IE8 support from our "Legacy" theme, which... we don't want. We want people who still depend on IE8 to have a solution in Backpack, even if we don't push any new stuff to it, right?

File: src/resources/views/base/inc/scripts.blade.php [hunk 4]
```diff
@@ -1,6 +1,10 @@
 @if (config('backpack.base.scripts') && count(config('backpack.base.scripts')))
     @foreach (config('backpack.base.scripts') as $path)
-    <script type="text/javascript" src="{{ asset($path).'?v='.config('backpack.base.cachebusting_string') }}"></script>
```
Comment (tabacitu) [inline src/resources/views/base/inc/scripts.blade.php:3]: Note that with this, we're no longer using the cachebusting string. Should we remove it from the config file? If we remove this, do we have a _different_ way of cachebusing, when using Bassets? 🐶

File: src/config/backpack/base.php (hunk 1 above)
Comment (tabacitu) [reply src/config/backpack/base.php:61]: See before (v5) and after (v6):

File: src/resources/views/base/widgets/style.blade.php [hunk 5]
```diff
@@ -6,9 +6,5 @@
 @endphp
 
 @push($widget['stack'] ?? 'after_styles')
-    <link href="{{ $href }}" type="text/css"
-        @foreach($attributes as $key => $value)
-        {{ $key }}{!! $value === true || $value === '' ? '' : "=\"$value\"" !!}
-        @endforeach
-    />
+    @basset($href, true, $attributes, 'style')
```
Comment (tabacitu) [inline src/resources/views/base/widgets/style.blade.php:9]: Beautiful!

perfection_cap

File: webpack.mix.js [hunk 6]
```diff
@@ -36,47 +36,7 @@ mix.sass('src/resources/assets/scss/blue-bundle.scss', 'src/public/packages/back
...
-	.copy('node_modules/datatables.net', 'src/public/packages/datatables.net')
-	.copy('node_modules/datatables.net-bs4', 'src/public/packages/datatables.net-bs4')
-	.copy('node_modules/datatables.net-fixedheader', 'src/public/packages/datatables.net-fixedheader')
-	.copy('node_modules/datatables.net-fixedheader-bs4', 'src/public/packages/datatables.net-fixedheader-bs4')
-	.copy('node_modules/datatables.net-responsive', 'src/public/packages/datatables.net-responsive')
-	.copy('node_modules/datatables.net-responsive-bs4', 'src/public/packages/datatables.net-responsive-bs4')
-	.copy('node_modules/places.js/dist', 'src/public/packages/places.js/dist')
-	.copy('node_modules/jquery-ui-touch-punch/jquery.ui.touch-punch.min.js', 'src/public/packages/jquery-ui-touch-punch/jquery.ui.touch-punch.min.js')
-	.copy('node_modules/urijs/src/URI.min.js', 'src/public/packages/URI.js/URI.min.js');
-
```
Comment (tabacitu) [inline webpack.mix.js:78]: yey_cap

**Review by tabacitu — CHANGES_REQUESTED** [review]
This is very VERY promising. Thanks a lot for doing the heavy lifting on this @promatik - I'm glad it's actually possible the way I imagined it 😅 Though... we do have quite a few things quirks to iron out. So let's do it, let's see if we can turn this into a 100% solution!

File: src/resources/views/crud/reorder.blade.php [hunk 7]
```diff
@@ -221,8 +221,8 @@ function tree_element($entry, $key, $all_entries, $crud)
 @endsection
 
 @section('after_scripts')
-    <script src="{{ asset('packages/jquery-ui-dist/jquery-ui.min.js') }}" type="text/javascript" ></script>
-    <script src="{{ asset('packages/nestedSortable/jquery.mjs.nestedSortable2.js') }}" type="text/javascript" ></script>
+    @basset('https://unpkg.com/jquery-ui@1.13.2/dist/jquery-ui.min.js')
+    @basset('https://cdnjs.cloudflare.com/ajax/libs/nestedSortable/2.0.0/jquery.mjs.nestedSortable.min.js')
```
Comment (tabacitu) [inline src/resources/views/crud/reorder.blade.php:225]: This isn't working for me. Once I drag something:

CleanShot 2023-01-03 at 22 04 10

File: src/config/backpack/base.php (hunk 1 above)
Comment (tabacitu) [reply src/config/backpack/base.php:61]: Hmm... there may be a simple solution for the first problem here - see https://github.com/DigitallyHappy/assets/pull/15#discussion_r1060946567

File: src/resources/views/base/inc/scripts.blade.php [hunk 8]
```diff
@@ -1,9 +1,20 @@
...
+        @endif
     @endforeach
 @endif
 
+@basset('https://cdnjs.cloudflare.com/ajax/libs/pace/1.2.4/pace.min.js')
+@basset('https://cdnjs.cloudflare.com/ajax/libs/noty/3.1.4/noty.min.js')
+@basset('https://cdnjs.cloudflare.com/ajax/libs/sweetalert/2.1.2/sweetalert.min.js')
+@basset('https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.3/jquery.min.js')
+@basset('https://cdnjs.cloudflare.com/ajax/libs/bootstrap/4.6.2/js/bootstrap.min.js')
+@basset('https://unpkg.com/@coreui/coreui@2.1.16/dist/js/coreui.js')
```
Comment (tabacitu) [inline src/resources/views/base/inc/scripts.blade.php:16]: Heads-up @promatik - this is important. What I've done here is that I've removed the `bundle.js` file from CRUD. It was just merging all of these together. So instead of needing that part of the build script... boom, the blade file is loading them. One less thing to build.

File: package.json [hunk 9]
```diff
@@ -17,49 +17,11 @@
...
-        "places.js": "^1.19.0",
-        "popper.js": "^1.16.1",
-        "select2": "^4.0.13",
-        "select2-bootstrap-theme": "0.1.0-beta.10",
-        "simple-line-icons": "^2.5.5",
-        "source-sans-pro": "^3.6",
-        "summernote": "^0.8.20",
-        "sweetalert": "^2.1.2",
-        "tinymce": "^5.10.2",
-        "urijs": "^1.19.11"
```
Comment (tabacitu) [inline package.json:63]: I haven't commited the `packages` directory itself, it would make the number of changed files EXPLODE. But look at how clean the `packages` directory is now, after you run `npm install && npm run prod` in CRUD:

CleanShot 2023-01-03 at 23 34 18

I've chosen to leave these 3, because:
- [ ] we're still bundling `backstrap` and the custom styles here together; but we should move these custom styles to `backstrap`, bundle it there, so we can then include it from CDN directly;
- [ ] we still haven't solved loading fonts from Google CDN or such; so right now we're loading them from local; but once we solve that, we should be able to remove the font from here too;

Comment (tabacitu) [reply package.json:63]: After we fix the two items above, we should be able to remove:
- `package.json` file
- `webpack.mix.js` file
- the `public/packages` directory entirely

That's when we have our first theme, the `base` theme.

File: src/config/backpack/base.php [hunk 10]
```diff
@@ -58,8 +58,12 @@
         // 'packages/noty/noty.css',
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/config/backpack/base.php]: Done! We're now using Backstrap 0.5.1, which also includes all of the customizations we previously had in CRUD.

File: package.json (hunk 9 above)
Comment (tabacitu) [reply package.json:63]: Done! Removed all of those files 🎉

File: src/resources/views/base/inc/head.blade.php (hunk 3 above)
Comment (tabacitu) [reply src/resources/views/base/inc/head.blade.php]: Didn't find a reason to remove this, so I've added it back.

**Comment by tabacitu** [comment]
Update: I've continued this branch, it's now as clean as it can get (without provoking huge changes in diffs). We've now removed ALL of the JS and CSS, and are only loading it using Bassets.

This does spark a few problems though. But we can consider them minor / solvable. See the inline comments above. After those are solved, we can do another round of testing... re-merge and close this 💪

File: src/resources/views/crud/reorder.blade.php (hunk 7 above)
Comment (tabacitu) [reply src/resources/views/crud/reorder.blade.php:225]: Fixed in https://github.com/Laravel-Backpack/CRUD/pull/4930

File: src/resources/views/base/inc/scripts.blade.php (hunk 4 above)
Comment (tabacitu) [reply src/resources/views/base/inc/scripts.blade.php:3]: Moved question to issue https://github.com/Laravel-Backpack/bassets/issues/6

**Comment by tabacitu** [comment]
Merged into the `v6` branch. We're done here 🎉

https://github.com/Laravel-Backpack/CRUD/pull/4860

**Comment by tabacitu** [comment]
Keeping this PR just for the diff.
